tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java

```diff
@@ -0,0 +1,124 @@
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import net.sourceforge.argparse4j.inf.Namespace;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * A consumer based EOS test handler that will poll data from input topic and
+ * buffer them in an in-memory queue.
+ */
+class ConsumerTestHandler<K, V> implements ClientTestHandler<K, V> {
+
+    private final Consumer<K, V> consumer;
+    private final String consumerGroupId;
+    private Deque<ProducerRecord<K, V>> recordsToBeSent;
+    private final long pollTimeout;
+    private final String inputTopic;
+    private final String outputTopic;
+    private final Producer<K, V> producer;
+
+    ConsumerTestHandler(Namespace res,
+                        String consumerGroupId,
+                        Producer<K, V> producer) throws IOException {
+        this.consumerGroupId = consumerGroupId;
+        List<String> consumerProps = res.getList("consumerConfig");
+        String consumerConfigFile = res.getString("consumerConfigFile");
+        pollTimeout = res.getLong("pollTimeout");
+        inputTopic = res.getString("inputTopic");
+        outputTopic = res.getString("topic");
+
+        Properties consumerProperties = KafkaClientPerformance.generateProps(consumerProps, consumerConfigFile);
+        consumerProperties.put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroupId);
+        consumerProperties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        consumerProperties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        this.consumer = new KafkaConsumer<>(consumerProperties);
+        this.consumer.subscribe(Collections.singletonList(inputTopic));
+
+        this.producer = producer;
+        recordsToBeSent = new ArrayDeque<>();
+    }
+
+    @Override
+    public ProducerRecord<K, V> getRecord() {
+        System.out.println("Consumer polls new data with timeout " + pollTimeout);
+        while (recordsToBeSent.isEmpty()) {
+            ConsumerRecords<K, V> consumedRecords = consumer.poll(Duration.ofMillis(pollTimeout));
```

Comment:
It looks like my comment below about getting a null message (as mentioned in the interface doc) would not happen because we'd spin here forever. Probably we should return null here if the fetch fails? Then in the loop below we may want to either fail immediately, because the test is going to be skewed or at least fail after n retries to avoid hanging everything, e.g. in the case of an undersized input topic.

Reply:
Yes, that's a good catch!